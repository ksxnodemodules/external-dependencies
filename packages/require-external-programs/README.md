# Require External Programs

Prevent user from installing your package without required executables

## Requirements

* Node.js ≥ 8.9.0

## Usage

Add this to your `package.json`:

```json
{
  "dependencies": {
    "require-external-programs": "<version range>"
  },
  "requireExternalPrograms": [
    "git",
    "python"
  ]
}
```

## Notes

* This package assumes you run `npm install` at the root of your project.

## License

[MIT](https://git.io/fx1N6) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
