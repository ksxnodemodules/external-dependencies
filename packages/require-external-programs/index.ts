export namespace scripts {
  export const install = require.resolve('./install')
}
