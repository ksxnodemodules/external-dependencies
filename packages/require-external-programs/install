#! /usr/bin/env node
require(require('require-external-programs-bin').bin)
