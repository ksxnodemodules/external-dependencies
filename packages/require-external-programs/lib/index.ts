export * from './types'
export * from './constants'
export * from './classes'
export * from './functions'
