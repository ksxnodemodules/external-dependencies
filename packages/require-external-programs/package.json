{
  "name": "require-external-programs",
  "version": "0.0.0",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/ksxnodemodules/external-dependencies#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodemodules/external-dependencies.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodemodules/external-dependencies/issues"
  },
  "dependencies": {
    "fs-extra": "^7.0.0",
    "glob": "^7.1.3",
    "glob-promise": "^3.4.0",
    "tslib": "^1.9.3",
    "@types/fs-extra": "^5.0.4",
    "@types/node": "^10.11.3"
  }
}
