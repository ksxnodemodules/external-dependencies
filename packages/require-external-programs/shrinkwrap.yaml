dependencies:
  '@types/fs-extra': 5.0.4
  '@types/node': 10.11.7
  advanced-set-mutable-multi-key: 0.0.1
  fs-extra: 7.0.0
  glob: 7.1.3
  glob-promise: 3.4.0
  tslib: 1.9.3
packages:
  /@babel/runtime/7.1.2:
    dependencies:
      regenerator-runtime: 0.12.1
    dev: false
    resolution:
      integrity: sha512-Y3SCjmhSupzFB6wcv1KmmFucH6gDVnI30WjOcicV10ju0cZjak3Jcs67YLIXBrmZYw1xCrVeJPbycFwrqNyxpg==
  /@types/events/1.2.0:
    dev: false
    resolution:
      integrity: sha512-KEIlhXnIutzKwRbQkGWb/I4HFqBuUykAdHgDED6xqwXJfONCjF5VoE0cXEiurh3XauygxzeDzgtXUqvLkxFzzA==
  /@types/fs-extra/5.0.4:
    dependencies:
      '@types/node': 10.11.7
    dev: false
    resolution:
      integrity: sha512-DsknoBvD8s+RFfSGjmERJ7ZOP1HI0UZRA3FSI+Zakhrc/Gy26YQsLI+m5V5DHxroHRJqCDLKJp7Hixn8zyaF7g==
  /@types/glob/7.1.1:
    dependencies:
      '@types/events': 1.2.0
      '@types/minimatch': 3.0.3
      '@types/node': 10.11.7
    dev: false
    resolution:
      integrity: sha512-1Bh06cbWJUHMC97acuD6UMG29nMt0Aqz1vF3guLfG+kHHJhy3AyohZFFxYk2f7Q1SQIrNwvncxAE0N/9s70F2w==
  /@types/minimatch/3.0.3:
    dev: false
    resolution:
      integrity: sha512-tHq6qdbT9U1IRSGf14CL0pUlULksvY9OZ+5eEgl1N7t+OA3tGvNpxJCzuKQlsNgCVwbAs670L1vcVQi8j9HjnA==
  /@types/node/10.11.7:
    dev: false
    resolution:
      integrity: sha512-yOxFfkN9xUFLyvWaeYj90mlqTJ41CsQzWKS3gXdOMOyPVacUsymejKxJ4/pMW7exouubuEeZLJawGgcNGYlTeg==
  /advanced-set-altered-equal/0.0.1:
    dependencies:
      '@types/node': 10.11.7
      advanced-set-base: 0.0.0
      advanced-set-types: 0.0.0
      convenient-typescript-utilities: 0.0.3
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-WvpNv4/zsZc6MYXPlpY0xGnjXXmWFWOUvc43eFfqKgSBWL6n5QLbkyneyjRdF6sBjDKXe2Yla4u1VsRt8IEdEQ==
  /advanced-set-base/0.0.0:
    dependencies:
      '@types/node': 10.11.7
      advanced-set-types: 0.0.0
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-RP3zt5zRkNTAWCW/0u1A3jmOp6xel8qFt0U9YYbuhNRGIkjGhHOb9lzchbND6yutVohaH32Voz0ojbOVIzkZLg==
  /advanced-set-multi-key-base/0.0.1:
    dependencies:
      '@types/node': 10.11.7
      advanced-set-altered-equal: 0.0.1
      advanced-set-types: 0.0.0
      create-array-equal: 0.0.0
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-kCZJkQZzpCbdmn830Bv6OX1MA9FlAuo3zJB0X6gdyPMqKSiA+lLk1JuBBEBjjT+XJdM1ISGMoUg70s16xOZTPg==
  /advanced-set-mutable-multi-key/0.0.1:
    dependencies:
      '@types/node': 10.11.7
      advanced-set-multi-key-base: 0.0.1
      advanced-set-types: 0.0.0
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-+9XfZs67S9r5Xh3wXNYDWZaspBHmKL1YA+NOhSsQSteRMrs6d8prPto0DB2bvXExkvlbiBG9Wo6pXwhQgVe2RA==
  /advanced-set-types/0.0.0:
    dependencies:
      '@types/node': 10.11.7
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-xAvd7khFkgWxmedqyBb53RW9JJwEOWahX/h2LaKx6YJkjHkriBPVkW1zX6K1RabUnWza6loqoKooe/oZkEQ+OQ==
  /balanced-match/1.0.0:
    dev: false
    resolution:
      integrity: sha1-ibTRmasr7kneFk6gK4nORi1xt2c=
  /brace-expansion/1.1.11:
    dependencies:
      balanced-match: 1.0.0
      concat-map: 0.0.1
    dev: false
    resolution:
      integrity: sha512-iCuPHDFgrHX7H2vEI/5xpz07zSHB00TpugqhmYtVmMO6518mCuRMoOYFldEBl0g187ufozdaHgWKcYFb61qGiA==
  /clone-regexp/2.1.0:
    dependencies:
      is-regexp: 2.0.0
    dev: false
    engines:
      node: '>=6'
    resolution:
      integrity: sha512-Ujn//SX3jg4duyQGM37swl+Sk95zmQnraZO0fGVZQEq1fJJML8mJaEkbT70ns/kBcdU7HLzpF8VJUmsb1lDD8Q==
  /concat-map/0.0.1:
    dev: false
    resolution:
      integrity: sha1-2Klr13/Wjfd5OnMDajug1UBdR3s=
  /convenient-typescript-utilities/0.0.3:
    dependencies:
      '@types/node': 10.11.7
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-kHjMz9faySaKMx9ZRwv9oTogYRXCGaKPYq+xsLEswBo+0a5r8l5Xk/7Q+vqDLMYMrTm2oW7wEgtiDcXonr8Nkg==
  /create-array-equal/0.0.0:
    dependencies:
      '@types/node': 10.11.7
      iter-tools: 6.1.4
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-PKGYy366uoR4/A1RqKgVmyZfxf7M7Ilp9L7zHx7OubCylYp4G2WcS3gnLwrfSKIUfbV8jOOYqY1YbNOhVGhR9g==
  /dequeue/1.0.5:
    dev: false
    resolution:
      integrity: sha1-EPHO8H4yNLIdyzj0v6LWYDSrZ8c=
  /fs-extra/7.0.0:
    dependencies:
      graceful-fs: 4.1.11
      jsonfile: 4.0.0
      universalify: 0.1.2
    dev: false
    engines:
      node: '>=6 <7 || >=8'
    resolution:
      integrity: sha512-EglNDLRpmaTWiD/qraZn6HREAEAHJcJOmxNEYwq6xeMKnVMAy3GUcFB+wXt2C6k4CNvB/mP1y/U3dzvKKj5OtQ==
  /fs.realpath/1.0.0:
    dev: false
    resolution:
      integrity: sha1-FQStJSMVjKpA20onh8sBQRmU6k8=
  /glob-promise/3.4.0:
    dependencies:
      '@types/glob': 7.1.1
    dev: false
    engines:
      node: '>=4'
    peerDependencies:
      glob: '*'
    resolution:
      integrity: sha512-q08RJ6O+eJn+dVanerAndJwIcumgbDdYiUT7zFQl3Wm1xD6fBKtah7H8ZJChj4wP+8C+QfeVy8xautR7rdmKEw==
  /glob/7.1.3:
    dependencies:
      fs.realpath: 1.0.0
      inflight: 1.0.6
      inherits: 2.0.3
      minimatch: 3.0.4
      once: 1.4.0
      path-is-absolute: 1.0.1
    dev: false
    resolution:
      integrity: sha512-vcfuiIxogLV4DlGBHIUOwI0IbrJ8HWPc4MU7HzviGeNho/UJDfi6B5p3sHeWIQ0KGIU0Jpxi5ZHxemQfLkkAwQ==
  /graceful-fs/4.1.11:
    dev: false
    engines:
      node: '>=0.4.0'
    resolution:
      integrity: sha1-Dovf5NHduIVNZOBOp8AOKgJuVlg=
  /inflight/1.0.6:
    dependencies:
      once: 1.4.0
      wrappy: 1.0.2
    dev: false
    resolution:
      integrity: sha1-Sb1jMdfQLQwJvJEKEHW6gWW1bfk=
  /inherits/2.0.3:
    dev: false
    resolution:
      integrity: sha1-Yzwsg+PaQqUC9SRmAiSA9CCCYd4=
  /is-regexp/2.0.0:
    dev: false
    engines:
      node: '>=6'
    resolution:
      integrity: sha512-B5NGyyvXTmppdR/g7ei6cu5DXhvzBLNcajOXJ3OR1tmW39iiF7444beL1lQjL393j8vDOTWGnSV2r/5cKuNqkg==
  /iter-tools/6.1.4:
    dependencies:
      '@babel/runtime': 7.1.2
      clone-regexp: 2.1.0
      dequeue: 1.0.5
      little-ds-toolkit: 1.1.0
      typescript-tuple: 1.7.3
      typescript-union: 0.0.0
    dev: false
    resolution:
      integrity: sha512-MpUCYuPa7w8CzoHFWRycVsWBookzztSYjFaeYWHBUmumtuNWKtAbhUnAS+9tEku9GA961a8xa7CeH5Vm0+pQ5g==
  /jsonfile/4.0.0:
    dev: false
    optionalDependencies:
      graceful-fs: 4.1.11
    resolution:
      integrity: sha1-h3Gq4HmbZAdrdmQPygWPnBDjPss=
  /little-ds-toolkit/1.1.0:
    dev: false
    resolution:
      integrity: sha512-2ARbmay5TrhFxFqcyFxPigmRToDuo8yxvIMdvwsDYNd/oZKnvsFT5L3I7s8WLXR7/tThYeffN9ouDtFFk4wzsA==
  /minimatch/3.0.4:
    dependencies:
      brace-expansion: 1.1.11
    dev: false
    resolution:
      integrity: sha512-yJHVQEhyqPLUTgt9B83PXu6W3rx4MvvHvSUvToogpwoGDOUQ+yDrR0HRot+yOCdCO7u4hX3pWft6kWBBcqh0UA==
  /once/1.4.0:
    dependencies:
      wrappy: 1.0.2
    dev: false
    resolution:
      integrity: sha1-WDsap3WWHUsROsF9nFC6753Xa9E=
  /path-is-absolute/1.0.1:
    dev: false
    engines:
      node: '>=0.10.0'
    resolution:
      integrity: sha1-F0uSaHNVNP+8es5r9TpanhtcX18=
  /regenerator-runtime/0.12.1:
    dev: false
    resolution:
      integrity: sha512-odxIc1/vDlo4iZcfXqRYFj0vpXFNoGdKMAUieAlFYO6m/nl5e9KR/beGf41z4a1FI+aQgtjhuaSlDxQ0hmkrHg==
  /tslib/1.9.3:
    dev: false
    resolution:
      integrity: sha512-4krF8scpejhaOgqzBEcGM7yDIEfi0/8+8zDRZhNZZ2kjmHJ4hv3zCbQWxoJGz1iw5U0Jl0nma13xzHXcncMavQ==
  /typescript-miscellaneous/0.1.0:
    dependencies:
      '@types/node': 10.11.7
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
      typescript: '>= 3.0.0'
    resolution:
      integrity: sha512-JPy2LNt44g/q4nJo4CzzGgfrA5LOHQl7I2XKVfK1gGz9F5WhataKRLrahk7RTMTckq31gkKe5LgiSkjxB2Ubiw==
  /typescript-tuple/1.7.3:
    dev: false
    engines:
      typescript: ^3.0.0
    resolution:
      integrity: sha512-ygCTfD6CX5ul+J0sZhq2MSouMNAeJpRXAFkYPUZomv457KWg15DuS7n+ubv15LtbX27iVo6DkuiX/5yyNdBvVg==
  /typescript-union/0.0.0:
    dependencies:
      typescript-tuple: 1.7.3
      utility-types: 2.1.0
    dev: false
    peerDependencies:
      typescript: '>= 3.0.0'
    resolution:
      integrity: sha512-sWOTNcTOXw27cJhfvjSbxrKggb4XD+fPIFkjIqqnycdVqStj/3IR3QVEMZOheWY5ELGS3PCwuHFDVWAbK9UHvA==
  /universalify/0.1.2:
    dev: false
    engines:
      node: '>= 4.0.0'
    resolution:
      integrity: sha512-rBJeI5CXAlmy1pV+617WB9J63U6XcazHHF2f2dbJix4XzpUF0RS3Zbj0FGIOCAva5P/d/GBOYaACQ1w+0azUkg==
  /utility-types/2.1.0:
    dev: false
    engines:
      node: '>= 4'
    resolution:
      integrity: sha512-/nP2gqavggo6l38rtQI/CdeV+2fmBGXVvHgj9kV2MAnms3TIi77Mz9BtapPFI0+GZQCqqom0vACQ+VlTTaCovw==
  /wrappy/1.0.2:
    dev: false
    resolution:
      integrity: sha1-tSQ9jz7BqjXxNkYFvA0QNuMKtp8=
registry: 'https://registry.npmjs.org/'
shrinkwrapMinorVersion: 9
shrinkwrapVersion: 3
specifiers:
  '@types/fs-extra': ^5.0.4
  '@types/node': ^10.11.3
  advanced-set-mutable-multi-key: ^0.0.1
  fs-extra: ^7.0.0
  glob: ^7.1.3
  glob-promise: ^3.4.0
  tslib: ^1.9.3
