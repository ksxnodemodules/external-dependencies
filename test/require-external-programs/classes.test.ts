import { Relationship } from 'require-external-programs'
class Instance extends Relationship<number, string> {}

function create () {
  const instance = new Instance()
  const snap = () => new Map(instance)
  return { instance, snap }
}

it('.add', () => {
  const { instance, snap } = create()
  const a = snap()
  const instance0 = instance.add(0, 'zero')
  const b = snap()
  const instance1 = instance0.add(1, 'one')
  const c = snap()

  expect({ a, b, c }).toMatchSnapshot()
  expect(instance0).toBe(instance)
  expect(instance1).toBe(instance)
})

it('.delete()', () => {
  const { instance, snap } = create()

  const a = snap()
  instance.add(0, 'zero').add(1, 'one').add(2, 'two')
  const b = snap()
  const del1one = instance.delete(1, 'one')
  const del3two = instance.delete(3, 'two')
  const del0two = instance.delete(0, 'two')
  const c = snap()

  expect({
    snapshots: { a, b, c },
    deletion: { del1one, del3two, del0two }
  }).toMatchSnapshot()
})

it('.has()', () => {
  const snap = () => [instance.has(0, 'zero'), instance.has(1, 'one')]
  const { instance } = create()

  const a = snap()
  instance.add(0, 'zero').add(1, 'one')
  const b = snap()
  instance.delete(0, 'zero')
  instance.add(0, '0')
  const c = snap()

  expect({ a, b, c }).toMatchSnapshot()
})
