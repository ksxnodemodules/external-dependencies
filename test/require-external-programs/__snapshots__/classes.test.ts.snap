// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relationship .add 1`] = `
Object {
  "a": Map {},
  "b": Map {
    0 => "zero",
  },
  "c": Map {
    0 => "zero",
    1 => "one",
  },
}
`;

exports[`Relationship .delete() 1`] = `
Object {
  "deletion": Object {
    "del0two": false,
    "del1one": true,
    "del3two": false,
  },
  "snapshots": Object {
    "a": Map {},
    "b": Map {
      0 => "zero",
      1 => "one",
      2 => "two",
    },
    "c": Map {
      0 => "zero",
      2 => "two",
    },
  },
}
`;

exports[`Relationship .has() 1`] = `
Object {
  "a": Array [
    false,
    false,
  ],
  "b": Array [
    true,
    true,
  ],
  "c": Array [
    false,
    true,
  ],
}
`;
